Implement finality certificates

Stebalien opened the pull request:
This implements finality certificates and power table deltas. This adds:

1. Methods for applying/making power table deltas.
2. A method for validating a chain of finality certificates.

Notes:

- Finality certificates include the power-table delta between the power table used to validate the current certificate and the next certificate. They don't include the delta implied by the chain head. We'll have to address this when we implement the finality certificate exchange protocol, but I expect we'll communicate these power-deltas separate from the certificates.

- I used `PowerEntries` instead of `PowerTable` objects because I doubt we'll need the performance. But I'm happy to change that.

fixes #5
fixes #14

Kubuxu commented:
One early note: it should live in a separate package. The finality certificate is quite separate from the gpbft itself.

Stebalien commented:
@Kubuxu I've moved it to a `certs` package, and renamed `certs` to `certstore`.

I now remember why I didn't do this: I was hoping to find a way to completely disconnect gpbft and finality certificates such that finality certificates didn't depend on gpbft. But... that's not critical and something we can look into later (gpbft itself doesn't have any external dependencies anyways).

Kubuxu commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+		}
+
+		// We can't have a finality certificate for "bottom"
+		if cert.ECChain.IsZero() {
```
Is that right? I'm not certain. I'm pretty sure the GPBFT can terminate with a bottom result.

Kubuxu replied:
Or will the justification be for the Base?

Stebalien replied:
From what I can tell, we cannot "decide" zero: https://github.com/filecoin-project/go-f3/blob/4f6464d9c33fae650918347bcdb06b5911922e5b/gpbft/gpbft.go#L430-L432

We can commit to zero to start another round, but we'll keep going until we decide on _at least_ the base.

Kubuxu reviewed (approved):
Other than that one comment. SGWM

anorth commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+
+	if justification.Vote.Value.IsZero() {
+		return nil, xerrors.Errorf("got a decision for bottom for instance %d", justification.Vote.Instance)
+	}
```
More preconditions you might check:
- Vote value is not empty, and does Validate()
- Signers is not empty?
- Supplemental Merkel hash is the right length?

Edit: I see some of these are captured below, but perhaps some of these easier, pure-syntax ones would be better earlier?

Stebalien replied:
I can but, IMO, the best solution is just to call `ValidateFinalityCertificates` on the result. It's unclear where to draw the line:

- Validate the signature? Requires the verifier and network name.
- Check the supplemental data? Requires signature validation.
- Check that we have enough power? Requires the power table.
- Validate the bitfield?
- Validate the power delta? Requires the original power table.

So instead I opted to validate information that _won't_ end up in the finality certificate (i.e., make sure the justification is actually for a decide). Well, that and the fact that we're voting on _something_.

Options are:

1. Separate out a "syntactic validation" function from `ValidateFinalityCertificate` in both places.
2. Call `ValidateFinalityCertificates` inside `NewFinalityCertificate` (need to pass a power table, network name, etc.).
3. Tell the user to call `ValidateFinalityCertificates`.

IMO, we should be calling `ValidateFinalityCertificates` regardless which reduces the usefulness of 1. However, I hesitate to go with 2 because it makes it impossible to construct a finality certificate from a justification without a verifier, etc.

anorth commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+// finalized, the instance of the first invalid finality certificate, and the power table that
+// should be used to validate that finality certificate, along with the error encountered.
+func ValidateFinalityCertificates(verifier gpbft.Verifier, network gpbft.NetworkName, prevPowerTable gpbft.PowerEntries, nextInstance uint64, base *gpbft.TipSet,
+	certs ...FinalityCertificate) (_nextInstance uint64, chain gpbft.ECChain, newPowerTable gpbft.PowerEntries, err error) {
```
Channelling @masih , consider introducing structs for these high arity parameter and return types.

anorth commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("failed to make power table CID for finality certificate for instance %d: %w", cert.GPBFTInstance, err)
+		}
+
+		if !bytes.Equal(cert.SupplementalData.PowerTable, powerTableCid) {
```
Nit: seeing this in use, perhaps we can make this name more obvious. E.g. SupplementalData.NextBasePowerTable? It's still tricky though...

Stebalien replied:
It's not really the next _base_ power table, it's the power table used to validate the next finality certificate. I can call it "next certificate power table" or "next instance power table"?

anorth commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+	}
+
+	signers := make([]gpbft.PubKey, 0, len(powerTable))
+	signerPower := new(gpbft.StoragePower)
```
Nit: signersPower ?

Stebalien replied:
Fixed in #292.

anorth commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+	for _, e := range oldPowerMap {
+		diff = append(diff, PowerTableDelta{
+			ParticipantID: e.ID,
+			PowerDelta:    new(gpbft.StoragePower).Neg(e.Power),
```
There's a subtle point here that a power delta that takes the result to zero is to be interpreted as a deletion. Deletion of the signing key isn't explicit here. I think it's worth documenting this interpretation, perhaps with the PowerTableDelta type. Or through the code with a function like `newPowerDeltaDelete(id, e.power)`

Stebalien replied:
Fixed in #292, #291.

anorth commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+			if diff.PowerDelta.Sign() != 0 {
+				pe.Power = new(gpbft.StoragePower).Add(diff.PowerDelta, pe.Power)
+				if pe.Power.Sign() == 0 {
+					continue
```
The interpretation of a deletion here is subtle, worth a comment.

Stebalien replied:
Fixed in #291.

anorth commented on certs/certs.go:
```diff
@@ -0,0 +1,293 @@
+				pe.PubKey = diff.SigningKey
+			}
+		}
+		newPowerTable = append(newPowerTable, pe)
```
I think it's worth validating here that the signing key is non-empty. A bad input could lead to an invalid power table.

Stebalien replied:
Fixed in #291.

anorth commented on test/certificate_test.go:
```diff
@@ -0,0 +1,387 @@
+package test
```
Would this be more canonically located as certs_test in the certs package? I'm curious why it's here.

Stebalien replied:
No good reason. IIRC, it was because:

1. The certificates used to live in the gpbft package.
2. I didn't want to import the simulation.

Or something like that.

I'll move it in #293 (where I'll also be adding more tests).

anorth reviewed (commented):
Nice.

The power table delta calculation and application is complex enough to deserve a little more unit testing IMO:
- power gets added to the right entry
- change in key is processed with/without a change in power
- sequences like deletion and re-addition of an entry (requiring key to be re-specified)
